Source/SwiftLintBuiltInRules/Rules/Metrics/NestingRule.swift

```diff
@@ -12,137 +13,171 @@ struct NestingRule: Rule {
         nonTriggeringExamples: NestingRuleExamples.nonTriggeringExamples,
         triggeringExamples: NestingRuleExamples.triggeringExamples
     )
+}
 
-    private let omittedStructureKinds = SwiftDeclarationKind.variableKinds
-        .union([.enumcase, .enumelement])
-        .map(SwiftStructureKind.declaration)
+private struct ValidationData {
+    private(set) var typeLevel: Int = -1
+    private(set) var functionLevel: Int = -1
+    private var declStack = Stack<any DeclSyntaxProtocol>()
 
-    private struct ValidationArgs {
-        var typeLevel: Int = -1
-        var functionLevel: Int = -1
-        var previousKind: SwiftStructureKind?
-        var violations: [StyleViolation] = []
+    var isFunction: Bool {
+        declStack.peek()?.is(FunctionDeclSyntax.self) ?? false
```

Comment:
The only fact to know from the top stack object is whether it is a function. So let's only store this information and avoid to deal with potentially expensive existential types.